lib/eth/contract.rb

```diff
@@ -0,0 +1,301 @@
+      @encoder = Eth::Contract::Encoder.new
+      @formatter = Eth::Contract::Formatter.new
+      @decoder = Eth::Contract::Decoder.new
+      @sender = client.default_account
+      @gas_limit = @client.gas_limit
+      @gas_price = @client.max_fee_per_gas
+      @max_fee_per_gas = @client.max_fee_per_gas
+      @max_priority_fee_per_gas = @client.max_priority_fee_per_gas
+    end
+
+    # Creates a contract wrapper.
+    def self.create(file: nil, client: nil, code: nil, abi: nil, address: nil, name: nil, contract_index: nil)
+      contract = nil
+      client = Eth::Client.create "/tmp/geth.ipc"
+      if File.exist?(file)
+        contracts = Eth::Contract::Initializer.new(file, client).build_all
+        raise "No contracts compiled" if contracts.empty?
+        if contract_index
+          contract = contracts[contract_index].class_object.new
+        else
+          contract = contracts.first.class_object.new
+        end
+      else
+        abi = abi.is_a?(String) ? JSON.parse(abi) : abi.map(&:deep_stringify_keys)
+        contract = Eth::Contract.new(name, code, abi, client)
+        contract.build
+        contract = contract.class_object.new
+      end
+      contract.address = address
+      contract
+    end
+
+    def address=(addr)
+      @address = addr
+      @events.each do |event|
+        event.set_address(addr)
+        event.set_client(@client)
+      end
+    end
+
+    def deploy_payload(params)
+      if @constructor_inputs.present?
+        raise ArgumentError, "Wrong number of arguments in a constructor" and return if params.length != @constructor_inputs.length
+      end
+      deploy_arguments = @encoder.encode_arguments(@constructor_inputs, params)
+      "0x" + @code + deploy_arguments
+    end
+
+    def deploy_args(params)
+      add_gas_options_args({from: sender, data: deploy_payload(params)})
+    end
+
+    def send_transaction(tx_args)
+        @client.eth_send_transaction(tx_args)["result"]
+    end
+
+    def send_raw_transaction(payload, to = nil)
+      Eth.configure { |c| c.chain_id = @client.net_version["result"].to_i }
+      @nonce ||= @client.get_nonce(key.address)
+      args = {
+        from: key.address,
+        value: 0,
+        data: payload,
+        nonce: @nonce,
+        gas_limit: gas_limit,
+        gas_price: gas_price
+      }
+      args[:to] = to if to
+      tx = Eth::Tx.new(args)
+      tx.sign key
+      @client.eth_send_raw_transaction(tx.hex)["result"]
+    end
+
+    def deploy(*params)
+      if key
+        tx = send_raw_transaction(deploy_payload(params))
+      else
+        tx = send_transaction(deploy_args(params))
+      end
+      tx_failed = tx.nil? || tx == "0x0000000000000000000000000000000000000000000000000000000000000000"
+      raise IOError, "Failed to deploy, did you unlock #{sender} account? Transaction hash: #{tx}" if tx_failed
+      @deployment = Eth::Contract::Deployment.new(tx, @client)
+    end
+
+    def deploy_and_wait(*params, **args, &block)
```

Comment:
All these functions should be implemented for the client not the contract. See transfer_and_wait.